flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java

```diff
@@ -90,6 +92,9 @@
 	@GuardedBy("lock")
 	private boolean disposed;
 
+	/** Whether to discard the useless state when retrieve local checkpoint state. */
+	private boolean retrieveWithDiscard = true;
```

Comment:
Why do we need this? Is there any case for not doing the cleanup?

Reply:
Aha, this is just for passing the existing test case in `TaskLocalStateStoreImplTest` ... 
```java
	private void checkStoredAsExpected(List<TaskStateSnapshot> history, int off, int len) throws Exception {
		for (int i = off; i < len; ++i) {
			TaskStateSnapshot expected = history.get(i);
			Assert.assertTrue(expected == taskLocalStateStore.retrieveLocalState(i));
			Mockito.verify(expected, Mockito.never()).discardState();
		}
	}
```

Reply:
Then there are two better options in my opinion, because the flag is pure boilerplate:

- Change the test to check what we are doing now, because that is what happens in the real use-case.
- Maybe even better: split the method `retrieveLocalState` further: one method for pruning, one package-private method that does all the pure retrieval, logging, and `null` transformation. In the old `retrieveLocalState`, do the cleanup first, then the pure retrieval/logging. Call the package private method in the test.

Maybe the test should then also just do both?

Reply:
Agreed and I prefer the second option.